pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -1509,6 +1515,23 @@ public CompletableFuture<List<String>> getListOfTopics(NamespaceName namespaceNa
         }
     }
 
+    public CompletableFuture<List<String>> getListOfUserTopics(NamespaceName namespaceName, Mode mode) {
+        String key = String.format("%s://%s", mode, namespaceName);
+        final MutableBoolean initializedByCurrentThread = new MutableBoolean();
+        CompletableFuture<List<String>> queryRes = inProgressQueryUserTopics.computeIfAbsent(key, k -> {
+            initializedByCurrentThread.setTrue();
+            return getListOfTopics(namespaceName, mode).thenApplyAsync(list -> {
+                return TopicList.filterSystemTopic(list);
+            }, pulsar.getExecutor());
```

Comment:
Why don't you invoke the `whenCompelete` here to remove the element? So that we don't need `initializedByCurrentThread`.

Reply:
- I think removing the value out of the function `map.computeIfAbsent` is more safety.
- and a `MutableBoolean` will not cost much

Reply:
We are not removing the value in the `computeIfAbsent`, right? We are removing it asynchronously.

Reply:
After discussing with @poorbarcode , calling the `whenComplete` in the `computeIfAbsent` is not safe because it may remove the value while in the same thread of `computeIfAbsent`. I'm +1 for this. Thanks @poorbarcode